File: 12-Heaps/cpp/01-implement-max-heap.cpp
class Solution {
    vector<int> heap;

public:
    void initializeHeap() { heap.clear(); }

    void insert(int key) {
        heap.push_back(key);
        siftUp(heap.size() - 1);
    }

    void changeKey(int index, int newVal) {
        heap[index] = newVal;
        siftUp(index);
        siftDown(index);
    }

    void extractMax() {
        int last = heap.size() - 1;
        heap[0] = heap[last];
        heap.pop_back();
        if (!isEmpty()) siftDown(0);
    }

    bool isEmpty() const { return heap.empty(); }
    int getMax() const { return heap[0]; }
    int heapSize() const { return heap.size(); }

private:
    void siftUp(int i) {
        while (i > 0) {
            int p = (i - 1) / 2;
            if (heap[i] > heap[p]) {
                swap(heap[i], heap[p]);
                i = p;
            } else break;
        }
    }

    void siftDown(int i) {
        int n = heap.size();
        while (2 * i + 1 < n) {
            int l = 2 * i + 1, r = 2 * i + 2, big = l;
            if (r < n && heap[r] > heap[l]) big = r;
            if (heap[big] > heap[i]) {
                swap(heap[i], heap[big]);
                i = big;
            } else break;
        }
    }
};

File: 12-Heaps/python/01_implement_max_heap.py
class Solution:
    def __init__(self):
        self.heap = []

    def initializeHeap(self):
        self.heap.clear()

    def insert(self, key: int):
        self.heap.append(key)
        self._sift_up(len(self.heap) - 1)

    def changeKey(self, index: int, newVal: int):
        self.heap[index] = newVal
        self._sift_up(index)
        self._sift_down(index)

    def extractMax(self):
        last = len(self.heap) - 1
        self.heap[0] = self.heap[last]
        self.heap.pop()
        if not self.isEmpty():
            self._sift_down(0)

    def isEmpty(self) -> bool:
        return len(self.heap) == 0

    def getMax(self) -> int:
        return self.heap[0]

    def heapSize(self) -> int:
        return len(self.heap)

    def _sift_up(self, i: int):
        while i > 0:
            p = (i - 1) // 2
            if self.heap[i] > self.heap[p]:
                self.heap[i], self.heap[p] = self.heap[p], self.heap[i]
                i = p
            else:
                break

    def _sift_down(self, i: int):
        n = len(self.heap)
        while 2 * i + 1 < n:
            l, r = 2 * i + 1, 2 * i + 2
            big = l
            if r < n and self.heap[r] > self.heap[l]:
                big = r
            if self.heap[big] > self.heap[i]:
                self.heap[i], self.heap[big] = self.heap[big], self.heap[i]
                i = big
            else:
                break
